common/src/main/java/io/netty/util/Recycler.java

```diff
@@ -108,6 +118,14 @@ protected Recycler(int maxCapacity) {
     }
 
     protected Recycler(int maxCapacity, int maxSharedCapacityFactor) {
+        this(maxCapacity, maxSharedCapacityFactor, RATIO);
+    }
+
+    protected Recycler(int maxCapacity, int maxSharedCapacityFactor, int ratio) {
+        if (ratio > 0x40000000) {
```

Comment:
consider creating a variant of `findNextPositivePowerOfTwo` which will enforce the bounds at runtime (but also leave the one that just does `assert` for other cases that already check the bounds).